_posts/2022-10-27-kani-bolero.md

```diff
@@ -0,0 +1,317 @@
+        0,
+        0,
+    ],
+    0,
+)
+
+Error: 
+panicked at 'index out of bounds: the len is 5 but the index is 5', src/main.rs:6:21
+```
+
+The fuzzing engine found an input that causes the expected out-of-bounds error: an array of 5 zeros and an index of zero. 
+
+Let's see if Kani finds the same issue:
+
+```bash
+$ cargo bolero test check_rev --engine kani
+# snip ...
+SUMMARY:
+ ** 1 of 337 failed (1 unreachable)
+Failed Checks: index out of bounds: the length is less than or equal to the given index
+ File: "/home/ubuntu/examples/demos/rev-bolero/src/main.rs", line 6, in nth_rev
+```
+
+It did! Now, let's fix the bug by updating the buggy line to:
+
+```rust
+        let rev_index = arr.len() - index - 1;
+```
+
+and re-run the fuzzer:
+
+```
+$ cargo bolero test check_rev
+# snip ...
+#3773   REDUCE cov: 135 ft: 138 corp: 18/282b lim: 52 exec/s: 0 rss: 43Mb L: 8/32 MS: 3 ChangeByte-ChangeBinInt-EraseBytes-
+#262144 pulse  cov: 135 ft: 138 corp: 18/282b lim: 2620 exec/s: 131072 rss: 68Mb
+#524288 pulse  cov: 135 ft: 138 corp: 18/282b lim: 4096 exec/s: 131072 rss: 93Mb
+#1048576        pulse  cov: 135 ft: 138 corp: 18/282b lim: 4096 exec/s: 116508 rss: 142Mb
+```
+
+The fuzzer will run for a while without finding any failures. We can now run Kani to *prove* the assertion in our harness:
+
+```
+$ cargo bolero test check_rev --engine kani
+# snip ...
+SUMMARY:
+ ** 0 of 338 failed (1 unreachable)
+
+VERIFICATION:- SUCCESSFUL
+Verification Time: 0.90419155s
+```
+
+and voila! Kani managed to prove that there none of the checks can fail for any values of the inputs.
+
+## If we have proof capability, do we still need fuzzing?
+
+One question that might arise is: is there any point in using fuzzing alongside with Kani?
+In other words, if we can prove a harness for all inputs with Kani, would we still want to use fuzzing?
+The answer is yes, due to a number of reasons. 
+
+First of all, the amount of code that Kani can analyze is typically much smaller than the code that fuzzers can handle.
+The reason is that, as explained in the [Comparison with other tools](https://model-checking.github.io/kani/tool-comparison.html#comparison-with-other-tools) section of the Kani documentation, Kani uses a technique called [*model checking*](https://en.wikipedia.org/wiki/Model_checking) in which the program is encoded as a mathematical formula that is solved using a special solver, called a [SAT solver](https://en.wikipedia.org/wiki/SAT_solver).
+The complexity of solving this formula grows exponentially with the size of the program.
+Thus, in practice, one needs to limit the size of the program for Kani, for example through reducing the size of inputs (e.g. vector length), or "[stubbing](https://en.wikipedia.org/wiki/Method_stub)" out some of the code.
+Due to this limitation, Kani is best suited for proving small pieces of code, while fuzzing, is best suited for end-to-end/integration testing.
+So in that sense, Kani and fuzzing are complementary.
```

Comment:
Apart from complexity, another reason is that Kani performs an analysis, but doesn't actually execute the code. This isn't the case with fuzzing, right?

Reply:
Replaced this with @danielsn's paragraph.